lib/puppet/pops/types/p_type_set_type.rb

```diff
@@ -350,15 +350,7 @@ def eql?(o)
   end
 
   def instance?(o, guard = nil)
-    # KEY_NAME_AUTHORITY, KEY_VERSION and Pcore::KEY_PCORE_URI seem to be optional, for example
-    # https://github.com/puppetlabs/puppet/blob/6.18.0/lib/puppet/pops/model/ast.rb#L4952-L4956,
-    if o.is_a?(Hash) &&
-       o.include?(KEY_NAME) &&
-       o.include?(Pcore::KEY_PCORE_VERSION) &&
-      super(o, guard)
```

Comment:
The original slow behavior was due to the `super` implementation defined in https://github.com/puppetlabs/puppet/blob/426e375a4bf0d08de1b84f0439f5259c6a415e11/lib/puppet/pops/types/p_meta_type.rb#L24-L26, which does a full type inference. Currently only `PTypeSetType` and `PObjectType` extend `PMetaType` and both of them override `instance?`. To avoid this problem in the future, I'm thinking `PMetaType#instance` should be removed, or it should raise a `NotImplementedError` so that we don't accidentally infer a large collection.

Reply:
PMetaType is abstract (in comments at least), so I went with raising NotImplementedError. That way if you extend the class in the future it's clear that method needs to be implemented.